### PR TITLE
[HCBS] Re-factor HCBS unit tests

### PR DESCRIPTION
**Pull request by rocio-desantiago**
### Description

- Where possible `fireEvent` replaced with `userEvent`
- Removed `data-testid` and `getByTestId` from everywhere but the parser `parsing.test.tsx`
- Replaced accessibility test with existing common test `testA11y`
- Updated tests to be more thorough and have more context than elements just being visible by adding using `getByText` and `getByRole` or `toHaveAttribute`

### Related ticket(s)
CMDCT-3895

---
### How to test
* Review code and verify updates are an improvement to tests
* All tests passing ✅ 

---
### Author checklist

- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [ ] I have updated relevant documentation, if necessary

### CODE REVIEW

File: services/ui-src/src/components/app/Error.test.tsx [hunk 1]
```diff
@@ -5,9 +5,11 @@ import { testA11y } from "utils/testing/commonTests";
...
 
 describe("<Error />", () => {
-  test("Check that Error page renders", () => {
-    const { getByTestId } = render(errorView);
-    expect(getByTestId("error-view")).toBeVisible();
+  test("Check that Error message renders", () => {
+    const { getByRole } = render(errorView);
+    expect(getByRole("heading", { level: 2 })).toHaveTextContent(
+      "Sorry! An error has occurred."
+    );
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/app/Error.test.tsx:12]: It's usually preferable to use `screen` for this exact reason: next time someone wants to add an assertion that uses a different `getBy` function, they won't also have to touch this line of code.

```suggestion
    render(errorView);
    expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent(
      "Sorry! An error has occurred."
    );
```

File: services/ui-src/src/components/cards/Card.test.tsx [hunk 2]
```diff
@@ -14,14 +14,8 @@ describe("Test Card", () => {
   });
 
   test("Card is visible", () => {
-    expect(screen.getByTestId("card")).toBeVisible();
+    expect(screen.getByText("Mock child component")).toBeVisible();
   });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/cards/Card.test.tsx:18]: Comment: I'm just really loving reading all of these little improvements ❤️

File: services/ui-src/src/components/layout/Footer.test.tsx [hunk 3]
```diff
@@ -1,41 +1,54 @@
...
+      expect(screen.getByRole("img", { name: "HCBS logo" })).toHaveAttribute(
+        "alt",
+        "HCBS logo"
+      );
+      expect(
+        screen.getByRole("img", {
+          name: "Medicaid.gov: Keeping America Healthy",
+        })
+      ).toHaveAttribute("alt", "Medicaid.gov: Keeping America Healthy");
+    });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Footer.test.tsx:37]: The `name` in the `getByRole` query for images actually searches for alt text already, so the `toHaveAttribute` assertion is redundant here.

```suggestion
    test("Images on footer are visible", () => {
      expect(
        screen.getByRole("img", {
          name: "Department of Health and Human Services, USA",
        })
      ).toBeVisible();
      expect(screen.getByRole("img", { name: "HCBS logo" })).toBeVisible();
      expect(
        screen.getByRole("img", {
          name: "Medicaid.gov: Keeping America Healthy",
        })
      ).toBeVisible();
    });
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 4]
```diff
@@ -16,44 +18,58 @@ describe("<Header />", () => {
...
+        "HCBS logo"
+      );
+      expect(screen.getByRole("img", { name: "Help" })).toHaveAttribute(
+        "alt",
+        "Help"
+      );
+      expect(screen.getByRole("img", { name: "Account" })).toHaveAttribute(
+        "alt",
+        "Account"
+      );
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:36]: Same suggestion here.

```suggestion
      // find img elements
      expect(screen.getByRole("img", { name: "HCBS logo" }))..toBeVisible()
      expect(screen.getByRole("img", { name: "Help" })).toBeVisible();
      expect(screen.getByRole("img", { name: "Account" })).toBeVisible();
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 5]
```diff
@@ -16,44 +18,58 @@ describe("<Header />", () => {
...
-
-      const logoutButton = screen.getByRole("img", { name: /Logout/i });
+    test("Logs out user", async () => {
+      // Find button
+      const menuButton = screen.getByRole("button", {
+        name: "my account",
+      });
+      await act(async () => {
+        await userEvent.click(menuButton);
+      });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:63]: This is a fun one! `userEvent` actually knows all about the React render cycle, so it already wraps calls in `act()` internally - we shouldn't need to worry about it at all.

Once again the Word of the Author is here: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#wrapping-things-in-act-unnecessarily

If the test starts giving you warnings when `act()` is removed, we can resolve those another way.

```suggestion
      await userEvent.click(menuButton);
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 6]
```diff
@@ -16,44 +18,58 @@ describe("<Header />", () => {
...
+        name: "my account",
+        expanded: false,
+      });
       expect(menuButton).toBeInTheDocument();
-
-      userEvent.click(menuButton);
+      // Click and expand
+      await act(async () => {
+        await userEvent.click(menuButton);
+      });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:50]: Same suggestion here (2/3)

```suggestion
      await userEvent.click(menuButton);
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 7]
```diff
@@ -16,44 +18,58 @@ describe("<Header />", () => {
...
+      // Click and expand
+      const logoutButton = screen.getByRole("img", { name: "Logout" });
       expect(logoutButton).toBeInTheDocument();
-
-      userEvent.click(logoutButton);
+      await act(async () => {
+        await userEvent.click(logoutButton);
+      });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:69]: And here 😎 (3/3)

```suggestion
      await userEvent.click(logoutButton);
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 8]
```diff
@@ -1,12 +1,14 @@
-import { render, screen } from "@testing-library/react";
+import { act, render, screen } from "@testing-library/react";
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:1]: And with the calls to `act()` gone, of course we won't need the import

```suggestion
import { render, screen } from "@testing-library/react";
```

File: services/ui-src/src/components/layout/Header.test.tsx [hunk 9]
```diff
@@ -16,44 +18,58 @@ describe("<Header />", () => {
...
+      // Click and expand
+      const logoutButton = screen.getByRole("img", { name: "Logout" });
       expect(logoutButton).toBeInTheDocument();
-
-      userEvent.click(logoutButton);
+      await act(async () => {
+        await userEvent.click(logoutButton);
+      });
+      // Mock logout
+      expect(mockLogout).toHaveBeenCalledTimes(1);
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/layout/Header.test.tsx:71]: This is a great test. I love the new assertion.

File: services/ui-src/src/components/pages/HelpPage/HelpPage.test.tsx [hunk 10]
```diff
@@ -16,14 +17,48 @@ describe("Test HelpPage", () => {
...
+  });
+
+  test("Check for email links", () => {
+    const email1 = screen.getByRole("link", { name: "mdct_help@cms.hhs.gov" });
+    expect(email1).toHaveAttribute("href", "mailto:mdct_help@cms.hhs.gov");
+    const email2 = screen.getByRole("link", { name: "MFPDemo@cms.hhs.gov" });
+    expect(email2).toHaveAttribute("href", "mailto:MFPDemo@cms.hhs.gov");
+  });
+
+  test("Question 1: When Expand button clicked it collapses", async () => {
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/pages/HelpPage/HelpPage.test.tsx:32]: I think this is a good test with the wrong name - if I'm reading it right, we're testing that it expands. Same with the test for Question 2.

There are also some `act()` calls in here that can probably be removed.

```suggestion
  test("Question 1: When Expand button clicked it expands", async () => {
```

File: services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx [hunk 11]
```diff
@@ -27,22 +29,25 @@ describe("Test ProfilePage for admin users", () => {
...
 
-  test("Check that admin button navigates to /admin on click", () => {
-    const adminButton = screen.getByTestId("banner-admin-button");
-    expect(adminButton).toBeVisible();
-    fireEvent.click(adminButton);
+  test("Check that admin button navigates to /admin on click", async () => {
+    await act(async () => {
+      const adminButtons = screen.getAllByRole("button");
+      await userEvent.click(adminButtons[0]);
+    });
```
Comment (benmartin-coforma) [inline services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx:50]: I think we can remove the `act()` here, but I think we can also improve the query. Hopefully this button has a unique `name` we can pass to `getByRole`?

**Review by benmartin-coforma — COMMENTED** [review]
This PR is beautiful. You're doing excellent work, Rocio.

File: services/ui-src/src/components/app/Error.test.tsx (hunk 1 above)
Comment (rocio-desantiago) [reply services/ui-src/src/components/app/Error.test.tsx:12]: Lol that makes sense. It originally said `screen` wasn't available and that was my workaround and I just realized, I just needed to import it 💀

File: services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx (hunk 11 above)
Comment (rocio-desantiago) [reply services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx:50]: It doesn't have a unique name. "Banner Editor" appears twice.

Comment (rocio-desantiago) [reply services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx:50]: removing 'act' causes the test to fail and it doesn't find the path "/admin"

Comment (benmartin-coforma) [reply services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx:50]: We have two buttons named Banner Editor?? That seems really odd 👀

Comment (rocio-desantiago) [reply services/ui-src/src/components/pages/Profile/ProfilePage.test.tsx]: I think it was a fluke. I updated the tests. It's not showing up twice anymore.

**Review by benmartin-coforma — DISMISSED** [review]

**Review by benmartin-coforma — APPROVED** [review]

**Review by ajaitasaini — APPROVED** [review]
code looks good 👍

**Pull request merged**